src/System/IO/Streams/SequenceId.hs

```diff
@@ -19,23 +20,28 @@ import qualified System.IO.Streams   as Streams
 -- Example:
 --
 -- @
--- ghci> 'System.IO.Streams.fromList' [1..10 :: 'Data.SequenceId.SequenceId'] >>= 'sequenceIdInputStream' 0 'id' ('fail' . 'show') >>= 'System.IO.Streams.toList'
--- [1,2,3,4,5,6,7,8,9,10]
---
--- ghci> 'System.IO.Streams.fromList' [5..10 :: 'Data.SequenceId.SequenceId'] >>= 'sequenceIdInputStream' 0 'id' ('fail' . 'show') >>= 'System.IO.Streams.toList'
--- *** Exception: user error ('Data.SequenceId.SequenceIdError' {errType = 'Data.SequenceId.SequenceIdDropped', lastSeqId = 0, currSeqId = 5})
---
--- ghci> 'System.IO.Streams.fromList' [1..10 :: 'Data.SequenceId.SequenceId'] >>= 'sequenceIdInputStream' 5 'id' ('fail' . 'show') >>= 'System.IO.Streams.toList'
--- *** Exception: user error ('Data.SequenceId.SequenceIdError' {errType = 'Data.SequenceId.SequenceIdDuplicated', lastSeqId = 5, currSeqId = 1})
+-- ghci> import qualified System.IO.Streams as Streams
+-- ghci> is <- Streams.fromList [1..10::Int]
+-- ghci> (is', resetSeqId) <- sequenceIdInputStream 0 id (fail . show) is
+-- ghci> Streams.read is'
+-- Just 1
+-- ghci> Streams.read is'
+-- Just 2
+-- ghci> Streams.read is'
+-- Just 3
+-- ghci> resetSeqId
+-- 3
+-- ghci> Streams.read is'
+-- *** Exception: user error (SequenceIdError {errType = SequenceIdDropped, lastSeqId = 0, currSeqId = 4})
 -- @
 sequenceIdInputStream :: Integral s
                       => s                            -- ^ Initial sequence ID
                       -> (a -> s)                     -- ^ Function applied to each element of the stream to get the sequence ID
                       -> (SequenceIdError s -> IO ()) -- ^ Error handler
                       -> InputStream a                -- ^ 'System.IO.Streams.InputStream' to check the sequence of
-                      -> IO (InputStream a)           -- ^ Pass-through of the given stream
-sequenceIdInputStream initSeqId getSeqId seqIdFaultHandler inStream =
-    fst <$> Streams.inputFoldM f initSeqId inStream
+                      -> IO (InputStream a, IO s)     -- ^ Pass-through of the given stream, and IO action to fetch and reset the seed value
```

Comment:
Is it possible to assert the sequence id via `inputFoldM`? Hmm.

I'm a little confused about the output of `IO s`. When you say "fetch and reset", does that mean fetch the new or the old value? The docs should make that clear.

Reply:
Unfortunately not. `IO s` here will return the current sequence id value (last seen, in this context) and then reset it to whatever the initial seed was. I can just make a new version that returns `s -> IO s`, even though I think in most cases we don't care about the current sequence id.

Reply:
It is not possible to set the sequence id to anything other than the initial value when using `inputFoldM`.  Another problem with using `inputFoldM` is that we may want setting/resetting to throw an error if the stream has failed due to a `SequenceIdError`.

Reply:
If you guys are curious, [here](https://github.com/bitnomial/seqid-streams/pull/4/commits/0b141c7a0e4512c0919b7faa7d4d0cea68edc180) is an earlier commit in which I hand wrote two functions that address Ian's concerns. It's less polished it since I abandoned them in favor of writing new code, but let me know if you think we should go this route instead.

Reply:
@bradsherman your work here is nice and clean.  I'm happy to stick with this PR.